### PR TITLE
Depot + Sub Versioner Refactor

### PR DESCRIPTION
**Pull request by Shikkic**
This PR allows us to use Depot in place of github!

Things to be done still
- [x] API end point for READMEs
- [x] Sub-versioner pushes to depot
- [x] Can create repos on volume programmatically
- [x] Check ref exists now uses depot
- [x] Web proxies depot
- [x] Sub-versioner now creates repos per repo+version (instead of version branches)
- [x] Public clone access from depot works as intended
- [ ] Needs to be optimized for prod still (aka use prod volume)
- [x] Router returns correct depot url
- [x] Tests for `download_package.go` @skeswa
- [x] Tests for `request_handler.go` @skeswa
- [ ] Tests for `record_package_download.go` @Shikkic 
- [ ] Tests for `record_package_archival.go` @Shikkic 
- [x] Tests for `push_to_depot.go` @Shikkic
- [ ] Tests for `workdir.go` @skeswa
- [ ] Tests for `delete_folder.go` @skeswa
- [ ] Tests for `errors.go`? (not sure on this one)
- [ ] Tests for `main.go`? (not sure on this one)
- [-] ~~Tests for API, Router, Sub Versioner~~ _I'm thinking that these can go in a different PR._

### CODE REVIEW

**Comment by Shikkic** [comment]
WOW! This took a bit of time but it's come together nicely. I just test this build and it works great locally.

@skeswa any chance you might be able to help me with the final two things? The only things left are:
- [ ] Needs to be optimized for prod still (aka use prod volume)
- [ ] Router returns depot url

I'm going to look at some of the router code tomorrow. I know a bit of it needs to be refactored, but I'm not sure exactly what needs to be refactored so that depot is used for cloning. We can discuss more tomorrow. As for now...

njueirh

File: api/handler_readme.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+		// Check if ref exists
+		if len(ref) < 1 {
+			errors.RespondWithError(w, NewInvalidQueryStringParameterError(
+				queryStringRefTextKey,
+				ref,
+			))
+			return
+		}
+
+		url := fmt.Sprintf("http://depot-svc:3000/?p=%s;a=blob_plain;f=README.md;hb=%s", repoName, ref)
```
Comment (skeswa) [inline api/handler_readme.go]: Pull `depot-svc` and `3000` out to constants. Could change in future.

File: api/handler_readme.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		resp, err := http.Get(url)
+		if err != nil {
+			errors.RespondWithError(w, err)
+			return
+		}
+		if resp.StatusCode == 404 {
+			w.WriteHeader(http.StatusNotFound)
+			w.Write([]byte{})
+		}
```
Comment (skeswa) [inline api/handler_readme.go]: Very few comments in this region of code. Consider revising that.

File: common/depot/check_ref.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+)
+
+const (
+	refsFetchURLTemplate          = "http://depot-svc:3000/%s.git/info/refs?service=git-upload-pack"
```
Comment (skeswa) [inline common/depot/check_ref.go]: See other comment about extracting constants.

File: common/depot/create_new_repo.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+const (
+	filePerm = 0644
+)
+
+// CreateNewRepo if repo doesn't already exist will create a new
+// repo in depot
+func CreateNewRepo(author string, repo string, ref string) error {
+	log.Println("Creating New Repo")
+	folderName := fmt.Sprintf(
```
Comment (skeswa) [inline common/depot/create_new_repo.go]: No need to multi-line this. Go doesn't have an 80-char rule.

File: common/depot/create_new_repo.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	err := os.Mkdir(
+		fmt.Sprintf("%s/%s", depotReposPath, folderName),
+		filePerm,
+	)
+
+	// TODO(Shikkic): If we can't make the repo folder bail, that means someone else is already versioning
+	if err != nil {
+		if checkIfRepoFolderExists(folderName); err != nil {
```
Comment (skeswa) [inline common/depot/create_new_repo.go]: Makes no sense. I think you meant `if err = checkIfRepoFolderExists(folderName); err != nil`

File: common/depot/create_new_repo.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+			return fmt.Errorf("Error, could not create folder or verify that it currently exists. %v", err)
+		}
+	}
+
+	// Git init bare
```
Comment (skeswa) [inline common/depot/create_new_repo.go]: Missing period after this comment. Also, this comment is not very helpful.

File: common/depot/create_new_repo.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	_, err = git.InitRepository(fmt.Sprintf("%s/%s", depotReposPath, folderName), true)
+	if err != nil {
+		return fmt.Errorf("Error, could not initialize new repository. %v", err)
+	}
+
+	return err
+}
+
+func checkIfRepoFolderExists(folderName string) error {
```
Comment (skeswa) [inline common/depot/create_new_repo.go]: Missing function comment.

File: common/depot/helpers.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+	"net/http"
+	"time"
+)
+
+const (
+	depotReposPath = "/repos"
+)
+
+var (
+	httpClient = &http.Client{Timeout: 10 * time.Second}
```
Comment (skeswa) [inline common/depot/helpers.go]: Where is this used?

File: common/depot/create_new_repo.go (hunk 7 above)
Comment (Shikkic) [inline common/depot/create_new_repo.go]: Only exported functions are required to have comments by the go-linter, no? Unless you're saying you want _all_ functions to have comments as a styling standard?

File: common/depot/helpers.go (hunk 8 above)
Comment (Shikkic) [inline common/depot/helpers.go]:
```
common/depot/helpers.go
14:    	httpClient = &http.Client{Timeout: 10 * time.Second}
```

File: common/depot/create_new_repo.go (hunk 5 above)
Comment (Shikkic) [inline common/depot/create_new_repo.go]: Yep I did, good catch

File: api/handler_readme.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+// ReadmeHandler creates an HTTP request handler that responds to fuzzy package
+// searches.
+func ReadmeHandler() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		// Get URL Query Params
+		qs := r.URL.Query()
+		if qs != nil {
+			// Get repo name
+			repoName = qs.Get(queryStringRepoTextKey)
```
Comment (Shikkic) [inline api/handler_readme.go]: I'd like to point out this route is a clone of `handler_search.go`, are you saying that handler_search would suffer from the same problems you're mentioning above and would require a refactor?

```
package main

import (
	"net/http"

	"github.com/gocql/gocql"
	"github.com/skeswa/gophr/common/dtos"
	"github.com/skeswa/gophr/common/errors"
	"github.com/skeswa/gophr/common/models"
)

const (
	queryStringSearchTextKey = "q"
)

// SearchHandler creates an HTTP request handler that responds to fuzzy package
// searches.
func SearchHandler(session *gocql.Session) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		var searchText string

		qs := r.URL.Query()
		if qs != nil {
			searchText = qs.Get(queryStringSearchTextKey)
		}

		if len(searchText) < 1 {
			errors.RespondWithError(w, NewInvalidQueryStringParameterError(
				queryStringSearchTextKey,
				searchText,
			))
			return
		}

		packageModels, err := models.FuzzySearchPackages(session, searchText)
		if err != nil {
			errors.RespondWithError(w, err)
			return
		}

		packageListDTO := dtos.NewPackageListDTOFromPackageModelList(packageModels)
		if len(packageListDTO) > 0 {
			json, err := packageListDTO.MarshalJSON()
			if err != nil {
				errors.RespondWithError(w, err)
				return
			}

			w.WriteHeader(http.StatusOK)
			w.Write(json)
		} else {
			w.WriteHeader(http.StatusOK)
			w.Write([]byte(dtos.EmptyListJSON))
		}
	}
}
```

File: api/handler_readme.go (hunk 1 above)
Comment (Shikkic) [inline api/handler_readme.go]: Makes sense

File: api/handler_readme.go (hunk 2 above)
Comment (Shikkic) [inline api/handler_readme.go]: Ok. Although I'm a bit unclear of our what our style guide should be? Some people are anti-comments (IE: Uncle Bob's clean code). Comments on exported functions makes sense for sure though. Thoughts?

File: api/handler_readme.go (hunk 9 above)
Comment (skeswa) [inline api/handler_readme.go]: Wat. The example from the search handler has all transient state _inside_ the closure.

File: api/handler_readme.go (hunk 2 above)
Comment (skeswa) [inline api/handler_readme.go]: Could just be me, but I like comments that outline intention. I have no idea what the purpose of each block written here. I would have to guess. This comment is much more on the optional side.

File: common/depot/create_new_repo.go (hunk 7 above)
Comment (skeswa) [inline common/depot/create_new_repo.go]: Oops. For some reason my brain said this was capitalized. My mistake.

File: api/handler_readme.go (hunk 9 above)
Comment (Shikkic) [inline api/handler_readme.go]: Whoops I see what you're saying. I can just move the variable declaration inside the closure and it should A-OK right?

File: common/depot/helpers.go (hunk 8 above)
Comment (skeswa) [inline common/depot/helpers.go]: Bruh. I know its declared where I put the comment. Where is it _used_. I couldn't see it used anywhere in this file.

File: api/handler_readme.go (hunk 9 above)
Comment (skeswa) [inline api/handler_readme.go]: Yup.

File: common/depot/helpers.go (hunk 8 above)
Comment (Shikkic) [inline common/depot/helpers.go]: Whoooops my bad,  I thought I linked to where it was used. It's used in `check_refs.go` line 23. Although I removed it as a global var and placed it inline.

Comment (skeswa) [inline common/depot/helpers.go]: Cool.

**Comment by Shikkic** [comment]
screen shot 2016-09-20 at 5 24 24 am

dat test coverage

**Comment by skeswa** [comment]
Ayyyyyyyyy

**Comment by Shikkic** [comment]
Is this ready to be merged?

**Comment by Shikkic** [comment]
- [ ] Write package_request.go test (that failed?) 
- [ ] CreateNewPackageRequest || respond
- [ ] Add prod volumes to kubernetes setup (Place them in controller && service.yml)
- [ ] Include git2go deps in indexer, api
-- Merge --
Gophrctl prod build api, router, indexer, web, depot

**Pull request merged**